### PR TITLE
Extend Serialization

### PR DESCRIPTION
**Pull request by mparno**
Should be merged after #351.
-  Adds methods for serializing TriangularMaps, MonotoneComponents, and everything they depend on (quadrature, polynomials, etc...)
- It's now possible to directly serialize a map using Cereal:
```cpp
std::shared_ptr<TriangularMap> tmap;
// Build tmap here...

// Serialize directly with Cereal
std::stringstream ss;
cereal::BinaryOutputArchive ar(ss);
ar(tmap);

// Deserialize directly with Cereal
std::shared_ptr<TriangularMap> tmap2;
cereal::BinaryInputArchive ar(ss);
ar(tmap2);
```
- It's also possible to use the `ParameterizedFunctionBase` class's `Save` and `Load` functions to save/load from any output stream.
```cpp
std::stringstream ss;
tmap->Save(ss);
tmap2 = ParameterizedFunctionBase<Kokkos::HostSpace>::Load(ss);
```
- This new functionality also allows the python bindings to be pickled natively:
```python
import pickle
import mpart as mt 

tmap = mt.TriangularMap(...)

bytes = pickle.dumps(tmap)
tmap2 = pickle.loads(bytes)
```

### CODE REVIEW

File: MParT/ConditionalMapBase.h [hunk 1]
```diff
@@ -149,13 +157,23 @@ namespace mpart {
...
+    // Define a serialize or save/load pair as you normally would
+    template <class Archive>
+    void save( Archive & ar ) const
+    {
+        ar( cereal::base_class<ParameterizedFunctionBase<MemorySpace>>( this )); 
+    }
+    template <class Archive>
+    void load( Archive & ar )
+    {
+        std::cout << "Shouldn't be here!" << std::endl;
```
Comment (dannys4) [inline MParT/ConditionalMapBase.h:172]: This should be clarified a little bit.

File: MParT/MonotoneComponent.h [hunk 2]
```diff
@@ -1023,4 +1073,26 @@ class MonotoneComponent : public ConditionalMapBase<MemorySpace>
 };
 
 } // namespace mpart
+
+#if defined(MPART_HAS_CEREAL)
+CEREAL_FORCE_DYNAMIC_INIT(mpart1)
```
Comment (dannys4) [inline MParT/MonotoneComponent.h:1078]: would I be a villain if I suggested we name these like `mpartInitMapFactory1`?

File: MParT/ParameterizedFunctionBase.h [hunk 3]
```diff
@@ -180,17 +186,49 @@ namespace mpart {
...
+    // Define a serialize or save/load pair as you normally would
+    template <class Archive>
+    void save( Archive & ar ) const
+    {
+        // Do nothing.
+    }
+    template <class Archive>
+    void load( Archive & ar )
+    {
+        std::cout << "Shouldn't be here!" << std::endl;
```
Comment (dannys4) [inline MParT/ParameterizedFunctionBase.h:229]: See above suggestion.

File: bindings/python/src/ConditionalMapBase.cpp [hunk 4]
```diff
@@ -23,6 +23,21 @@ void mpart::binding::ConditionalMapBaseWrapper(py::module &m)
...
+            },
+            [](py::bytes input) {
+                
+                std::stringstream ss;
+                ss.str(input);
+
+                auto ptr = std::dynamic_pointer_cast<ConditionalMapBase<Kokkos::HostSpace>>(ParameterizedFunctionBase<Kokkos::HostSpace>::Load(ss));
+                return ptr;
+            }
+        ));
```
Comment (dannys4) [inline bindings/python/src/ConditionalMapBase.cpp:40]: Super pedantic, but remove semicolon at end

File: bindings/python/src/MultiIndex.cpp [hunk 5]
```diff
@@ -214,6 +214,25 @@ void mpart::binding::MultiIndexWrapper(py::module &m)
...
+                FixedMultiIndexSet<Kokkos::HostSpace> mset;
+
+                std::stringstream ss;
+                ss.str(input);
+                cereal::BinaryInputArchive iarchive(ss);
+                iarchive(mset);
+
+                return mset;
+            }
+        ));
```
Comment (dannys4) [inline bindings/python/src/MultiIndex.cpp:235]: Semicolon

File: bindings/python/src/ParameterizedFunctionBase.cpp [hunk 6]
```diff
@@ -26,13 +26,40 @@ void mpart::binding::ParameterizedFunctionBaseWrapper<Kokkos::HostSpace>(py::mod
...
+            },
+            [](py::bytes input) {
+                
+                std::stringstream ss;
+                ss.str(input);
+
+                auto ptr = ParameterizedFunctionBase<Kokkos::HostSpace>::Load(ss);
+                return ptr;
+            }
+        ));
```
Comment (dannys4) [inline bindings/python/src/ParameterizedFunctionBase.cpp:62]: Semicolon

File: bindings/python/src/TriangularMap.cpp [hunk 7]
```diff
@@ -17,9 +17,24 @@ void mpart::binding::TriangularMapWrapper(py::module &m)
...
+            },
+            [](py::bytes input) {
+                
+                std::stringstream ss;
+                ss.str(input);
+
+                auto ptr = std::dynamic_pointer_cast<TriangularMap<Kokkos::HostSpace>>(ParameterizedFunctionBase<Kokkos::HostSpace>::Load(ss));
+                return ptr;
+            }
+        ));
```
Comment (dannys4) [inline bindings/python/src/TriangularMap.cpp:37]: Semicolon

**Review by dannys4 — CHANGES_REQUESTED** [review]
Mostly good, a few styling questions. Sorry for a bunch of comments that just say "semicolon"-- I've had my fair share of issues not recognizing someone put an auxilliary semicolon in python bindings, and I'm just trying to keep consistent design patterns on this.

**Comment by dannys4** [comment]
Also looks like it fails one of the tests? I can't immediately understand why.

File: MParT/MonotoneComponent.h (hunk 2 above)
Comment (mparno) [reply MParT/MonotoneComponent.h:1078]: That's fine.  We can actually put any name we want here as long as it matches what's in the header file.  It also does nothing but produce some code that causes the file to be tracked by Cereal.

**Review by dannys4 — APPROVED** [review]
As long as this passes tests, it looks good!

**Pull request merged**